File: examples/Erlang/anpr/src/anpr.erl
%% The MIT License (MIT)
%% Copyright © 2022-2025 TS-Solution Corp.

-module(anpr).
-export([main/0, read_image_file/4, read_encoded_image/4, read_pixel_buffer/4]).

-define(EXAMPLES_BASE_DIR, "../..").

get_engine_file_name() ->
    case os:type() of
        {win32, _} ->
            filename:join([?EXAMPLES_BASE_DIR, "bin", "windows-x86_64", "tsanpr.dll"]);
        {unix, linux} ->
            case erlang:system_info(system_architecture) of
                "aarch64" ++ _ ->
                    filename:join([?EXAMPLES_BASE_DIR, "bin", "linux-aarch64", "libtsanpr.so"]);
                _ ->
                    filename:join([?EXAMPLES_BASE_DIR, "bin", "linux-x86_64", "libtsanpr.so"])
            end;
        _ ->
            ""
    end.

%% Read image file using engine
read_image_file(TSANPR, ImgFile, OutputFormat, Options) ->
    io:format("~s (outputFormat=\"~s\", options=\"~s\") => ~n", [ImgFile, OutputFormat, Options]),
    Result = tsanpr:anpr_read_file(TSANPR, ImgFile, OutputFormat, Options),
    io:format("~s~n", [Result]).

%% Read encoded image as a buffer
read_encoded_image(TSANPR, ImgFile, OutputFormat, Options) ->
    io:format("~s (outputFormat=\"~s\", options=\"~s\") => ~n", [ImgFile, OutputFormat, Options]),
    case file:read_file(ImgFile) of
        {ok, EncodedImg} ->
            Result = tsanpr:anpr_read_pixels(TSANPR, EncodedImg, byte_size(EncodedImg), 0, 0, "encoded", OutputFormat, Options),
            io:format("~s~n", [Result]);
        {error, Reason} ->
            io:format("File read failed: ~p~n", [Reason])
    end.

%% Read decoded pixel buffer (simplified implementation for demonstration)
read_pixel_buffer(_TSANPR, ImgFile, OutputFormat, Options) ->
    io:format("~s (outputFormat=\"~s\", options=\"~s\") => ~n", [ImgFile, OutputFormat, Options]),
    io:format("Pixel buffer processing not implemented in this example~n").

%% Recognize license plates for a given country
read_license_plates(TSANPR, CountryCode) ->
    %% NOTICE:
    %% anpr_initialize should be called only once after library load.
    %% Therefore, it is not possible to change the country code after anpr_initialize has been called.
    %% While using the free trial license, you can try all languages.
    %% When you purchase a commercial license, you can only use the selected language.
    InitParams = "text;country=" ++ CountryCode,
    case tsanpr:anpr_initialize(TSANPR, InitParams) of
        "" ->
            ImageDir = filename:join([?EXAMPLES_BASE_DIR, "img", CountryCode]),

            %% TODO: Try each function as needed
            ANPRFunc = fun read_image_file/4,
            %% ANPRFunc = fun read_encoded_image/4,
            %% ANPRFunc = fun read_pixel_buffer/4,

            %% TODO: Try each output format as needed
            OutputFormat = "text",
            %% OutputFormat = "json",
            %% OutputFormat = "yaml",
            %% OutputFormat = "xml",
            %% OutputFormat = "csv",

            ANPRFunc(TSANPR, filename:join(ImageDir, "licensePlate.jpg"), OutputFormat, ""),    %% Single license plate recognition (default)
            ANPRFunc(TSANPR, filename:join(ImageDir, "multiple.jpg"), OutputFormat, "vm"),      %% Recognize multiple license plates attached to vehicles
            ANPRFunc(TSANPR, filename:join(ImageDir, "multiple.jpg"), OutputFormat, "vmb"),     %% Recognize multiple license plates attached to vehicles (including motorcycles)
            ANPRFunc(TSANPR, filename:join(ImageDir, "surround.jpg"), OutputFormat, "vms"),     %% Recognize multiple license plates attached to vehicles with surround detection
            ANPRFunc(TSANPR, filename:join(ImageDir, "surround.jpg"), OutputFormat, "dms"),     %% Recognize multiple surrounding objects (vehicles)
            ANPRFunc(TSANPR, filename:join(ImageDir, "surround.jpg"), OutputFormat, "dmsr"),    %% Recognize multiple surrounding objects (vehicles) and license plates

            %% Recognize multiple surrounding objects and license plates within RoI
            ANPRFunc(TSANPR, filename:join(ImageDir, "surround.jpg"), OutputFormat, "dmsri549,700,549,2427,1289,2427,1289,700");
        Error ->
            io:format("anpr_initialize() failed: ~s~n", [Error])
    end.

main() ->
    EngineFileName = get_engine_file_name(),
    case filelib:is_file(EngineFileName) of
        false ->
            io:format("Unsupported operating system or engine not found~n");
        true ->
            case tsanpr:load(EngineFileName) of
                {ok, TSANPR} ->
                    %% TODO: Try each country code as needed
                    read_license_plates(TSANPR, "KR");
                    %% read_license_plates(TSANPR, "JP");
                    %% read_license_plates(TSANPR, "VN");
                {error, Reason} ->
                    io:format("TSANPR initialization failed: ~p~n", [Reason])
            end
    end.

File: examples/Elixir/anpr/lib/anpr.ex
# The MIT License (MIT)
# Copyright © 2022-2025 TS-Solution Corp.

defmodule ANPR do
  @moduledoc """
  Main ANPR example module for Elixir.
  """

  alias TSANPR

  @examples_base_dir Path.expand("../../..", __DIR__)

  def get_engine_file_name do
    # Generate engine filename depending on platform and architecture.
    case :os.type() do
      {:win32, _} ->
        case :erlang.system_info(:wordsize) do
          8 -> Path.join([@examples_base_dir, "bin", "windows-x86_64", "tsanpr.dll"])
          _ -> Path.join([@examples_base_dir, "bin", "windows-x86", "tsanpr.dll"])
        end

      {:unix, :linux} ->
        case :erlang.system_info(:system_architecture) do
          ~c"x86_64" ++ _ -> Path.join([@examples_base_dir, "bin", "linux-x86_64", "libtsanpr.so"])
          ~c"aarch64" ++ _ -> Path.join([@examples_base_dir, "bin", "linux-aarch64", "libtsanpr.so"])
          _ -> ""
        end

      _ ->
        ""
    end
  end

  def read_image_file(tsanpr, imgfile, output_format, options) do
    # Read an image file and call anpr_read_file.
    IO.write("#{imgfile} (outputFormat=\"#{output_format}\", options=\"#{options}\") => ")
    result = TSANPR.anpr_read_file(tsanpr, imgfile, output_format, options)
    IO.puts(result)
  end

  def read_encoded_image(tsanpr, imgfile, output_format, options) do
    # Read an encoded image file as bytes and call TSANPR.anpr_read_pixels with 'encoded' pixel format.
    IO.write("#{imgfile} (outputFormat=\"#{output_format}\", options=\"#{options}\") => ")

    try do
      case File.exists?(imgfile) do
        false ->
          IO.puts("File does not exist")

        true ->
          {:ok, encoded_img} = File.read(imgfile)

          result =
            TSANPR.anpr_read_pixels(
              tsanpr,
              encoded_img,
              byte_size(encoded_img),
              0,
              0,
              "encoded",
              output_format,
              options
            )

          IO.puts(result)
      end
    rescue
      e -> IO.puts("ERROR: Exception - #{Exception.message(e)}")
    end
  end

  def read_pixel_buffer(_tsanpr, imgfile, output_format, options) do
    # Use the pixel buffer-based ANPR function.
    # Note: This is a simplified version. In a real implementation,
    # you would need to integrate with an image processing library.
    IO.write("#{imgfile} (outputFormat=\"#{output_format}\", options=\"#{options}\") => ")
    IO.puts("Pixel buffer reading not implemented in this example")
  end

  def read_license_plates(tsanpr, country_code) do
    # NOTICE:
    # anpr_initialize should be called only once after library load.
    # Therefore, it is not possible to change the country code after anpr_initialize has been called.
    # While using the free trial license, you can try all languages.
    # When you purchase a commercial license, you can only use the selected language.
    error = TSANPR.anpr_initialize(tsanpr, "text;country=#{country_code}")

    case error do
      "" -> :ok
      _ ->
        IO.puts("anpr_initialize() failed: #{error}")
        :error
    end

    image_dir = Path.join([@examples_base_dir, "img", country_code])

    # TODO: Try each function as needed
    anpr_func = &read_image_file/4
    # anpr_func = &read_encoded_image/4
    # anpr_func = &read_pixel_buffer/4

    # TODO: Try each output format as needed
    output_format = "text"
    # output_format = "json"
    # output_format = "yaml"
    # output_format = "xml"
    # output_format = "csv"

    # Single license plate recognition (default)
    anpr_func.(tsanpr, Path.join(image_dir, "licensePlate.jpg"), output_format, "")
    # Recognize multiple license plates attached to vehicles
    anpr_func.(tsanpr, Path.join(image_dir, "multiple.jpg"), output_format, "vm")
    # Recognize multiple license plates attached to vehicles (including motorcycles)
    anpr_func.(tsanpr, Path.join(image_dir, "multiple.jpg"), output_format, "vmb")
    # Recognize multiple license plates attached to vehicles with surround detection
    anpr_func.(tsanpr, Path.join(image_dir, "surround.jpg"), output_format, "vms")
    # Recognize multiple surrounding objects (vehicles)
    anpr_func.(tsanpr, Path.join(image_dir, "surround.jpg"), output_format, "dms")
    # Recognize multiple surrounding objects (vehicles) and license plates
    anpr_func.(tsanpr, Path.join(image_dir, "surround.jpg"), output_format, "dmsr")

    # Recognize multiple surrounding objects and license plates within RoI
    anpr_func.(
      tsanpr,
      Path.join(image_dir, "surround.jpg"),
      output_format,
      "dmsri549,700,549,2427,1289,2427,1289,700"
    )
  end

  def main do
    engine_file_name = get_engine_file_name()

    case {engine_file_name, File.exists?(engine_file_name)} do
      {"", _} ->
        IO.puts("Unsupported operating system or engine not found")

      {_, false} ->
        IO.puts("Unsupported operating system or engine not found")

      {engine_file, true} ->
        try do
          {:ok, tsanpr} = TSANPR.load(engine_file)

          # TODO: Try each country code as needed
          read_license_plates(tsanpr, "KR")
          # read_license_plates(tsanpr, "JP")
          # read_license_plates(tsanpr, "VN")
        rescue
          e -> IO.puts("TSANPR initialization failed: #{Exception.message(e)}")
        end
    end
  end
end
